server/src/main/java/org/apache/iotdb/db/trigger/executor/TriggerFireVisitor.java

```diff
@@ -0,0 +1,375 @@
+  }
+
+  @Override
+  public TriggerFireResult visitInsertMultiTablets(
+      InsertMultiTabletsNode node, TriggerEvent context) {
+    boolean hasFailedTrigger = false;
+    for (InsertTabletNode insertTabletNode : node.getInsertTabletNodeList()) {
+      TriggerFireResult result = visitInsertTablet(insertTabletNode, context);
+      if (result.equals(TriggerFireResult.TERMINATION)) {
+        return result;
+      }
+      if (result.equals(TriggerFireResult.FAILED_NO_TERMINATION)) {
+        hasFailedTrigger = true;
+      }
+    }
+    return hasFailedTrigger ? TriggerFireResult.FAILED_NO_TERMINATION : TriggerFireResult.SUCCESS;
+  }
+
+  @Override
+  public TriggerFireResult visitInsertRowsOfOneDevice(
+      InsertRowsOfOneDeviceNode node, TriggerEvent context) {
+    boolean hasFailedTrigger = false;
+    for (InsertRowNode insertRowNode : node.getInsertRowNodeList()) {
+      TriggerFireResult result = visitInsertRow(insertRowNode, context);
+      if (result.equals(TriggerFireResult.TERMINATION)) {
+        return result;
+      }
+      if (result.equals(TriggerFireResult.FAILED_NO_TERMINATION)) {
+        hasFailedTrigger = true;
+      }
+    }
+    return hasFailedTrigger ? TriggerFireResult.FAILED_NO_TERMINATION : TriggerFireResult.SUCCESS;
+  }
+
+  private Map<String, Integer> constructMeasurementToSchemaIndexMap(
+      String[] measurements, MeasurementSchema[] schemas) {
+    // The index of measurement and schema is the same now.
+    // However, in case one day the order changes, we need to construct an index map.
+    Map<String, Integer> indexMap = new HashMap<>();
+    for (int i = 0, n = measurements.length; i < n; i++) {
+      if (measurements[i] == null) {
+        continue;
+      }
+      // It is the same now
+      if (schemas[i] != null && schemas[i].getMeasurementId().equals(measurements[i])) {
+        indexMap.put(measurements[i], i);
+        continue;
+      }
+      for (int j = 0, m = schemas.length; j < m; j++) {
+        if (schemas[j] != null && schemas[j].getMeasurementId().equals(measurements[i])) {
+          indexMap.put(measurements[i], j);
+          break;
+        }
+      }
+    }
+    return indexMap;
+  }
+
+  private Map<String, List<String>> constructTriggerNameToMeasurementListMap(InsertNode node) {
+    Map<String, List<String>> triggerNameToPaths = new HashMap<>();
+    PartialPath device = node.getDevicePath();
+    if (!TriggerManagementService.getInstance().hasTriggerUnderDevice(device)) {
+      return triggerNameToPaths;
+    }
+    String[] measurements = node.getMeasurements();
+    for (String measurement : measurements) {
+      if (measurement != null) {
+        List<String> triggerList = getMatchedTriggerListForPath(device.concatNode(measurement));
+        for (String trigger : triggerList) {
+          triggerNameToPaths.computeIfAbsent(trigger, k -> new ArrayList<>()).add(measurement);
+        }
```

Comment:
You can make it more effective in next pr after PathPatternTreeMap offers more functions, because we don't need to search from scratch again, we should save all the nodes in the previous `device phase` and start from those nodes again for the remaining measurements.